composition_polynomial/src/periodic_columns.rs

```diff
@@ -0,0 +1,199 @@
+use ark_ff::PrimeField;
+use ark_poly::{EvaluationDomain, Polynomial, Radix2EvaluationDomain};
+use fri::lde::MultiplicativeLDE;
+
+#[derive(Debug, Clone)]
+pub struct PeriodicColumn<F: PrimeField> {
+    group_generator: F,
+    column_step: usize,
+    period_in_trace: usize,
+    n_copies: usize,
+    offset_compensation: F,
+    lde_manager: MultiplicativeLDE<F>,
+}
+
+impl<F: PrimeField> PeriodicColumn<F> {
+    pub fn new(
+        values: Vec<F>,
+        group_generator: F,
+        offset: F,
+        coset_size: usize,
+        column_step: usize,
+    ) -> Self {
+        let period_in_trace = values.len() * column_step;
+
+        assert!(period_in_trace != 0);
+        assert!(coset_size % period_in_trace == 0);
+        let n_copies = coset_size / period_in_trace;
+
+        let offset_compensation = offset.pow([n_copies as u64]).inverse().unwrap();
+
+        let base = Radix2EvaluationDomain::new_coset(values.len(), F::ONE).unwrap();
+        let mut lde_manager = MultiplicativeLDE::new(base, false);
+        lde_manager.add_eval(&values);
+
+        Self {
+            group_generator,
+            column_step,
+            period_in_trace,
+            n_copies,
+            offset_compensation,
+            lde_manager,
+        }
+    }
+
+    pub fn eval_at_point(&self, x: F) -> F {
```

Comment:
Why do we need the `lde_manager`?

We have `values` and assuming the length of the table divides the group order the polynomial for the column has a succinct description (and can be evaluated efficiently). So can't we just evaluate it directly?

For instance, the polynomial f(X) which evaluates to:

```
f(1)   = y0
f(w)   = y1
f(w^2) = y0
f(w^3) = y1
...
```

i.e. alternating. For a domain of length $2^k$, is simply:

$$
f(X) = a \cdot X^{2^k / 2} + b
$$

Where $a, b$: 

$$
f(1) = a + b = y0
$$

$$
f(w) = a + b * w^{2^k / 2} = a - b = y1
$$

So $a = (y0 + y1) / 2$ and $b = (y0 - y1) / 2$. A similar deal applies when you have a "starting_point", except now the polynomial is composed with * starting_point.


Here is a Sage script if you want to play around it:

```sage
F = GF(2^251 + 17 * 2^192 + 1)
k = 192

g = F.multiplicative_generator()
g = g^(g.multiplicative_order() / 2^k)

order = 2^8
w = g^(g.multiplicative_order() / order)

print(f"order: {order}")
print(g.multiplicative_order())

table = [
    1, 2, 1, 2
]

ys = table * (order // len(table))
xs = [w^i for i in range(order)]

R.<X> = PolynomialRing(F)

f = R.lagrange_polynomial(zip(xs, ys))

print(f)
```

Otherwise, let's hop on a call.